adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -119,6 +118,480 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
+     *                the number of partitions requested for the join differs from the <this.partitions>.
+     * @param genomicRdd The RDD to join to.
+     * @return A tuple containing both RDDs ready to join. _._1 is the leftRdd and _._2 is the rightRdd
+     */
+    private def prepareForShuffleRegionJoin[X, Y <: GenomicRDD[X, Y], Z <: GenomicRDD[(T, X), Z]](leftRdd: GenomicRDD[T, U], genomicRdd: GenomicRDD[X, Y])(implicit tTag: ClassTag[T], xTag: ClassTag[X]): ((RDD[((ReferenceRegion, Int), T)], RDD[((ReferenceRegion, Int), X)])) = {
+      assert(leftRdd.SortedTrait.isSorted)
+      val rightRdd: GenomicRDD[X, Y] =
+        // if both are sorted, great! let's just co-partition the data properly
+        if (genomicRdd.SortedTrait.isSorted) {
+          genomicRdd.SortedTrait.coPartitionByGenomicRegion(leftRdd)
+        } // otherwise we will co-partition then sort
+        else {
+          genomicRdd.coPartitionByGenomicRegion(leftRdd)
+        }
+      assert(rightRdd.SortedTrait.isSorted)
+
+      val leftPmap = leftRdd.SortedTrait.partitionMap
+      val leftRddReadyToJoin = leftRdd.rdd.mapPartitionsWithIndex((idx, iter) => {
+        leftRdd.SortedTrait.inferCorrectReferenceRegionsForPartition(iter, leftPmap(idx)).map(f => ((f._1, idx), f._2))
+      })
+
+      val rightPmap = rightRdd.SortedTrait.partitionMap
+      val rightRddReadyToJoin = rightRdd.rdd.mapPartitionsWithIndex((idx, iter) => {
+        rightRdd.SortedTrait.inferCorrectReferenceRegionsForPartition(iter, rightPmap(idx)).map(f => ((f._1, idx), f._2))
+      })
+      (leftRddReadyToJoin, rightRddReadyToJoin)
+    }
+
+    /**
+     * Performs a sort-merge inner join between this RDD and another RDD.
+     *
+     * In a sort-merge join, both RDDs are co-partitioned and sorted. The
+     * partitions are then zipped, and we do a merge join on each partition.
+     * The key equality function used for this join is the reference region
+     * overlap function. Since this is an inner join, all values who do not
+     * overlap a value from the other RDD are dropped.
+     *
+     * @param genomicRdd The right RDD in the join.
+     * @param optPartitions An Option to give the number of partitions you would like
+     * @return Returns a new genomic RDD containing all pairs of keys that
+     *   overlapped in the genomic coordinate space.
+     */
+    def shuffleRegionJoin[X, Y <: GenomicRDD[X, Y], Z <: GenomicRDD[(T, X), Z]](genomicRdd: GenomicRDD[X, Y],
+                                                                                optPartitions: Option[Int] = None)(
+                                                                                  implicit tTag: ClassTag[T], xTag: ClassTag[X]): GenomicRDD[(T, X), Z] = {
+      // did the user provide a set partition count?
+      // if no, we will avoid repartitioning this one
+      val partitions = optPartitions.getOrElse(rdd.partitions.length)
```

Comment:
Since this code is used across all of the `shuffleRegionJoin` impl's let's factor it out into a helper function.